File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
    if (s.length % 2 == 1)
        return false

    let stack = []
    const bracketMap = {
        "]": "[",
        "}": "{",
        ")": "("
    }
    for (let i = 0; i < s.length; i++) {
        switch (s[i]) {
            case "[":
            case "{":
            case "(":
                stack.push(s[i])
                break;
            default:
                if (stack.pop() != bracketMap[s[i]])
                    return false;
                break;
        }
    }
    return stack.length == 0;
};

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> map = new HashMap<>();
        map.put(')', '(');
        map.put('}', '{');
        map.put(']', '[');
        Stack stack = new Stack<Character>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else if (stack.empty()) {
                return false;
            } else if (!stack.pop().equals(map.get(c))) {
                return false;
            }
        }
        return stack.empty();
    }
}